src/contracts/core/AllocationManager.sol

```diff
@@ -0,0 +1,733 @@
+                // or before the current timestamp: the number of pending allocations is simply length
+                if (currentMagnitude == 0 && pos == 0) {
+                    numPendingAllocations = length;
+                } else {
+                    // if lookup found, then we take the difference between length-1 and pos
+                    numPendingAllocations = length - 1 - pos;
+                }
+
+                uint256 numPendingDeallocations =
+                    _getNumQueuedDeallocations(operator, allocation.strategy, operatorSetKey);
+
+                require(
+                    numPendingAllocations + numPendingDeallocations < MAX_PENDING_UPDATES,
+                    PendingAllocationOrDeallocation()
+                );
+            }
+
+            if (allocation.magnitudes[i] < uint64(currentMagnitude)) {
+                // Newly configured magnitude is less than current value.
+                // Therefore we handle this as a deallocation
+
+                // Note: MAX_PENDING_UPDATES == 1, so we do not have to decrement any allocations
+
+                // 1. push PendingFreeMagnitude and respective array index into (op,opSet,Strategy) queued deallocations
+                uint64 magnitudeToDeallocate = uint64(currentMagnitude) - allocation.magnitudes[i];
+                _queuedDeallocationIndices[operator][allocation.strategy][operatorSetKey].push(
+                    _pendingFreeMagnitude[operator][allocation.strategy].length
+                );
+                _pendingFreeMagnitude[operator][allocation.strategy].push(
+                    PendingFreeMagnitude({
+                        magnitudeDiff: magnitudeToDeallocate,
+                        completableTimestamp: deallocationCompletableTimestamp
+                    })
+                );
+
+                // 2. decrement allocated magnitude
+                magnitudeUpdates.decrementAtAndFutureSnapshots({
+                    key: uint32(block.timestamp),
+                    decrementValue: magnitudeToDeallocate
+                });
+            } else if (allocation.magnitudes[i] > uint64(currentMagnitude)) {
+                // Newly configured magnitude is greater than current value.
+                // Therefore we handle this as an allocation
+
+                // 1. decrement free magnitude by incremented amount
+                uint64 magnitudeToAllocate = allocation.magnitudes[i] - uint64(currentMagnitude);
+                OperatorMagnitudeInfo storage info = operatorMagnitudeInfo[operator][allocation.strategy];
+                require(info.freeMagnitude >= magnitudeToAllocate, InsufficientAllocatableMagnitude());
+                info.freeMagnitude -= magnitudeToAllocate;
+
+                // 2. allocate magnitude which will take effect in the future 21 days from now
+                magnitudeUpdates.push({key: allocationEffectTimestamp, value: allocation.magnitudes[i]});
```

Comment:
nit on the comment here: referring to `21 days` when this should probably be referring to a defined constant instead